### PR TITLE
Fixed SVG-as-text image comparison tests.

### PR DESCRIPTION
**Pull request by tfpf**
SVG-as-text tests make make the figure background transparent, but its colour (i.e. under the transparent mask) is actually black. Before the expected and actual images are compared, they are converted from RGBA to RGB, which makes them both completely black (since the text is also black). As a result, the test passes even if the expected and actual images differ. Making the text white solves this problem.

## PR Summary
Originally noted in https://github.com/matplotlib/matplotlib/pull/22852#discussion_r906632550. If `$-$-` in `svgastext_math_tests` is changed to something else, the test still passes, because the colour of the transparent figure patch is actually black.

Ideally, the colour of the patch should be changed from black to white (instead of setting the colour of the text to white and changing the two images), but even after doing:
- `mpl.rcParams['figure.facecolor'] = 'white'`
- `mpl.rcParams['savefig.facecolor'] = 'white'`
- `fig.patch.set(visible=False, facecolor='white')`

the patch remained black. (Is there any other way?)

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by anntzer** [comment]
I can't repro the problem locally: the background gets converted to white, not to black, in my case.
In any case, it would seem that a better fix would be for the testing machinery to not ignore the alpha channel if it is present, no?  Does something like the following work for you?
```
diff --git i/lib/matplotlib/testing/compare.py w/lib/matplotlib/testing/compare.py
index 665e055f12..1119cf10f8 100644
--- i/lib/matplotlib/testing/compare.py
+++ w/lib/matplotlib/testing/compare.py
@@ -371,6 +371,11 @@ def calculate_rms(expected_image, actual_image):
 # 16-bit depth, as Pillow converts these to RGB incorrectly.
 
 
+def _load_image(path):
+    img = Image.open(path)
+    return np.asarray(img if img.mode == "RGBA" else img.convert("RGB"))
+
+
 def compare_images(expected, actual, tol, in_decorator=False):
     """
     Compare two "image" files checking differences within a tolerance.
@@ -435,9 +440,9 @@ def compare_images(expected, actual, tol, in_decorator=False):
         actual = convert(actual, cache=True)
         expected = convert(expected, cache=True)
 
-    # open the image files and remove the alpha channel (if it exists)
-    expected_image = np.asarray(Image.open(expected).convert("RGB"))
-    actual_image = np.asarray(Image.open(actual).convert("RGB"))
+    # open the image files
+    expected_image = _load_image(expected)
+    actual_image = _load_image(actual)
 
     actual_image, expected_image = crop_to_same(
         actual, actual_image, expected, expected_image)
@@ -486,9 +491,8 @@ def save_diff_image(expected, actual, output):
     output : str
         File path to save difference image to.
     """
-    # Drop alpha channels, similarly to compare_images.
-    expected_image = np.asarray(Image.open(expected).convert("RGB"))
-    actual_image = np.asarray(Image.open(actual).convert("RGB"))
+    expected_image = _load_image(expected)
+    actual_image = _load_image(actual)
     actual_image, expected_image = crop_to_same(
         actual, actual_image, expected, expected_image)
     expected_image = np.array(expected_image).astype(float)
```

**Comment by tfpf** [comment]
That's odd. It's noteworthy that in https://github.com/matplotlib/matplotlib/pull/22852, the latest commit contains two wrong images (one each for Computer Modern and DejaVu Sans), and the tests still passed on CI!

I'll try with these suggested changes, but perhaps it is more important to reproduce the problem first?

**Comment by tfpf** [comment]
After https://github.com/matplotlib/matplotlib/pull/23350/commits/d23db61f3d4a66160cf980d2a944b9b4327c7356, all checks (other than PR cleanliness) have passed, which demonstrates the issue.

Was there a good reason the transparency was not being considered?

@anntzer I'm not sure if I should be the one to make your suggested changes to this PR, since it was your idea and effort!:sweat_smile:

**Comment by anntzer** [comment]
It's OK, I donate the patch above to you if it works, feel free to give it a try.  (Sorry, I may not have the time to investigate the repro for now, although I agree it would be nice to figure out why it only occurs sometimes.)
My guess is that back when the test machinery was introduced, transparency support was more brittle and therefore skipped?  Haven't checked the git history, though.

**Comment by tfpf** [comment]
Got it; I'll take a look at the failures and check.

**Comment by tfpf** [comment]
I updated the function like this.
```
def _load_image(path):
    img = Image.open(path)
    if img.mode == "RGBA" and img.getextrema()[3][0] < 255:
        return np.asarray(img)
    return np.asarray(img.convert("RGB"))
```

One test (`test_bbox_inches_tight_suptile_legend[svg]`) fails with the error: `Image sizes do not match expected size: (569, 885, 4) actual size (569, 885, 3)`. This is a genuine error—not due to these changes, but one which was already present. It was not detected earlier because the alpha channel was ignored. Here's what their alpha channels look like.



wrong_alpha_in_exp_image

The actual image is completely opaque; the expected image isn't—its rightmost column has transparency (note the value reported at the mouse position).

**Comment by tfpf** [comment]
This transparency problem is a lot more complicated than I thought! On the current `main` branch, this is what I get when I run `test_bbox_inches_tight_suptile_legend[svg]`.
- `bbox_inches_tight_suptile_legend-expected.svg` has a transparent strip at its right edge.
- `bbox_inches_tight_suptile_legend.svg` also has a transparent strip at its right edge, but it is thinner.
- `bbox_inches_tight_suptile_legend_svg.png` does not have the transparent strip.

Questions:
- Why does the actual SVG have transparency? As far as I can see, the test doesn't do anything to make that happen.
- Why does the actual SVG-converted-to-PNG have no transparency?

@tacaswell A visual inspection of the history of this image shows that that latest commit f15c26b0a79a42441df89428e6c21037bba7f932 might have introduced the transparent strip. Could you throw some light here?:sweat_smile:

**Comment by tfpf** [comment]
After manually converting the actual and expected SVG images (using Inkscape) to PNG, I believe that the original image should not have a transparent strip.

File: lib/matplotlib/testing/compare.py [hunk 1]
```diff
@@ -371,6 +371,13 @@ def calculate_rms(expected_image, actual_image):
 # 16-bit depth, as Pillow converts these to RGB incorrectly.
 
 
+def _load_image(path):
+    img = Image.open(path)
+    if img.mode != "RGBA" or img.getextrema()[3][0] == 255:
```
Comment (QuLogic) [inline lib/matplotlib/testing/compare.py:376]: This is checking that alpha is all opaque? I think it could use a comment.

**Review by QuLogic — APPROVED** [review]

**Comment by tacaswell** [comment]
My very vague memory (refreshed from looking at #9817) is that after changing from ints -> floats in the svg without any other changes.  If it affected anything with transparency that was unintentional.

**Comment by tacaswell** [comment]
My guess is that the source of the strip is related to rounding / snapping on the clip box.

It is oddly re-assuring how much our SVGs have changed while still producing the same visual result!

**Comment by tacaswell** [comment]
Thank you @tfpf !

**Pull request merged**